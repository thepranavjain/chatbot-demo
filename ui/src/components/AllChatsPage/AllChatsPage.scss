.chat-session {
  .delete-button {
    position: absolute;
    top: 0;
    right: 12px;
    background: none;
    border: none;
    border-radius: 8px;
    display: none;
  }

  &:hover .delete-button {
    display: grid;
    place-content: center;
  }
}

.chat-session-card {
  text-decoration: none;
  color: inherit;
}

.disabled-link {
  pointer-events: none;
}