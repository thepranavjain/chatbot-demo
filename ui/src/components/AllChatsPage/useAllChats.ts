import { useState, useEffect } from "react";
import { ChatSession } from "../../utils/types";
import { axiosAuthInstance } from "../../utils/api";

const useAllChats = () => {
  const [chatSessions, setChatSessions] = useState<ChatSession[]>([]);

  useEffect(() => {
    const fetchChatSessions = async () => {
      try {
        const response = await axiosAuthInstance.get("/messaging/chat-session");
        setChatSessions(response.data);
      } catch (error) {
        console.error("Error fetching chat sessions:", error);
      }
    };

    fetchChatSessions();
  }, []);

  return { chatSessions };
};

export default useAllChats;
