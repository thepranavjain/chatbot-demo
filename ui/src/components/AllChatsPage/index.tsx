import React from "react";
import "./AllChatsPage.scss";
import useAllChats from "./useAllChats";
import { Link } from "react-router";
import { ReactComponent as DeleteIcon } from "../../assets/icons/close-icon.svg";
import LoadingSpinner from "../LoadingSpinner";
import { cx } from "@emotion/css";

const AllChatsPage: React.FC = () => {
  const { chatSessions, deleteChat, isApiCallOngoing, loading } = useAllChats();

  return (
    <div className="container py-3">
      <Link to={"/chat/new"}>
        <button className="btn btn-primary mb-3">+ New Chat</button>
      </Link>
      {loading ? (
        <LoadingSpinner />
      ) : (
        <div className="row">
          {chatSessions.map((session) => (
            <div
              className="chat-session col-12 col-md-4 mb-3 position-relative"
              key={session.id}
            >
              <Link
                to={`/chat/${session.id}`}
                className={cx("card chat-session-card", {
                  "disabled-link": isApiCallOngoing,
                })}
                style={{ textDecoration: "none", color: "inherit" }}
              >
                <div className="card-body">
                  <h5 className="card-title">{session.name}</h5>
                  <p className="card-text">
                    {new Date(session.created).toDateString()}
                  </p>
                </div>
              </Link>
              <button
                onClick={() => deleteChat(session.id)}
                className="btn btn-light delete-button p-1"
                disabled={isApiCallOngoing}
                data-testid={`delete-chat-${session.id}`}
              >
                <DeleteIcon height="20px" className="delete-icon" />
              </button>
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

export default AllChatsPage;
