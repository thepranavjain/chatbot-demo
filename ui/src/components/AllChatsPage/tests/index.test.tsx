import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import AllChatsPage from "../index";
import useAllChats from "../useAllChats";

jest.mock("../useAllChats");

describe("AllChatsPage", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (useAllChats as jest.Mock).mockReturnValue({
      chatSessions: [],
      deleteChat: jest.fn(),
      isApiCallOngoing: false,
      loading: false,
    });
  });

  it("renders loading spinner when loading", () => {
    (useAllChats as jest.Mock).mockReturnValueOnce({
      chatSessions: [],
      deleteChat: jest.fn(),
      isApiCallOngoing: false,
      loading: true,
    });

    const { asFragment } = render(
      <MemoryRouter>
        <AllChatsPage />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders chat sessions", () => {
    (useAllChats as jest.Mock).mockReturnValueOnce({
      chatSessions: [
        {
          id: 1,
          name: "Chat 1",
          user_email: "test@example.com",
          created: "2025-01-01T00:00:00Z",
          updated: "2025-01-01T00:00:00Z",
        },
        {
          id: 2,
          name: "Chat 2",
          user_email: "test@example.com",
          created: "2025-01-02T00:00:00Z",
          updated: "2025-01-02T00:00:00Z",
        },
      ],
      deleteChat: jest.fn(),
      isApiCallOngoing: false,
      loading: false,
    });

    const { asFragment } = render(
      <MemoryRouter>
        <AllChatsPage />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("calls deleteChat when delete button is clicked", () => {
    const deleteChatMock = jest.fn();
    (useAllChats as jest.Mock).mockReturnValueOnce({
      chatSessions: [
        {
          id: 1,
          name: "Chat 1",
          user_email: "test@example.com",
          created: "2025-01-01T00:00:00Z",
          updated: "2025-01-01T00:00:00Z",
        },
      ],
      deleteChat: deleteChatMock,
      isApiCallOngoing: false,
      loading: false,
    });

    render(
      <MemoryRouter>
        <AllChatsPage />
      </MemoryRouter>
    );

    const deleteButton = screen.getByTestId("delete-chat-1");
    fireEvent.click(deleteButton);

    expect(deleteChatMock).toHaveBeenCalledWith(1);
  });

  it("disables delete button when API call is ongoing", () => {
    (useAllChats as jest.Mock).mockReturnValueOnce({
      chatSessions: [
        {
          id: 1,
          name: "Chat 1",
          user_email: "test@example.com",
          created: "2025-01-01T00:00:00Z",
          updated: "2025-01-01T00:00:00Z",
        },
      ],
      deleteChat: jest.fn(),
      isApiCallOngoing: true,
      loading: false,
    });

    render(
      <MemoryRouter>
        <AllChatsPage />
      </MemoryRouter>
    );

    const deleteButton = screen.getByTestId("delete-chat-1");
    expect(deleteButton).toBeDisabled();
  });
});
