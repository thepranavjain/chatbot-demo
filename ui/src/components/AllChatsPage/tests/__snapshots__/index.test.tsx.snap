// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllChatsPage renders chat sessions 1`] = `
<DocumentFragment>
  <div
    class="container py-3"
  >
    <a
      data-discover="true"
      href="/chat/new"
    >
      <button
        class="btn btn-primary mb-3"
      >
        + New Chat
      </button>
    </a>
    <div
      class="row"
    >
      <div
        class="chat-session col-12 col-md-4 mb-3 position-relative"
      >
        <a
          class="card chat-session-card"
          data-discover="true"
          href="/chat/1"
          style="text-decoration: none;"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              Chat 1
            </h5>
            <p
              class="card-text"
            >
              Wed Jan 01 2025
            </p>
          </div>
        </a>
        <button
          class="btn btn-light delete-button p-1"
          data-testid="delete-chat-1"
        >
          <svg
            class="delete-icon"
            height="20px"
          >
            close-icon.svg
          </svg>
        </button>
      </div>
      <div
        class="chat-session col-12 col-md-4 mb-3 position-relative"
      >
        <a
          class="card chat-session-card"
          data-discover="true"
          href="/chat/2"
          style="text-decoration: none;"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              Chat 2
            </h5>
            <p
              class="card-text"
            >
              Thu Jan 02 2025
            </p>
          </div>
        </a>
        <button
          class="btn btn-light delete-button p-1"
          data-testid="delete-chat-2"
        >
          <svg
            class="delete-icon"
            height="20px"
          >
            close-icon.svg
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AllChatsPage renders loading spinner when loading 1`] = `
<DocumentFragment>
  <div
    class="container py-3"
  >
    <a
      data-discover="true"
      href="/chat/new"
    >
      <button
        class="btn btn-primary mb-3"
      >
        + New Chat
      </button>
    </a>
    <div
      class="spinner"
    >
      <svg
        class="loader-icon"
        height="40px"
        width="40px"
      >
        loader.svg
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
